compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs

```diff
@@ -440,66 +482,155 @@ genForBind :: (GenPhase ph, MonadEnv m ph)
   -- ^ The binding being bound with this update.
   -> Expr
   -- ^ The expression in which this binding is being made available.
-  -> m (Output ph, Type, Maybe Expr)
+  -> m (Output ph, Maybe Type, Maybe Expr)
 genForBind bind body = do
   bindOut <- genExpr False (bindingBound bind)
-  bindUpd <- case _oExpr bindOut of
+  (bindUpd, subst) <- case _oExpr bindOut of
     EUpdate (UFetch tc cid) -> do
-      bindCids (TContractId (TCon tc)) cid (EVar $ fst $ bindingBinder bind) Nothing
-      return emptyUpdateSet
+      let var0 = fst $ bindingBinder bind
+      var1 <- genRenamedVar var0
+      let subst = singleExprSubst var0 (EVar var1)
+      _ <- bindCids False (Just $ TContractId (TCon tc)) cid (EVar var1) Nothing
+      return (emptyUpdateSet, subst)
     EUpdate upd -> do
       (updOut, updTyp, creFs) <- genUpdate upd
       this <- genRenamedVar (ExprVarName "this")
-      bindCids updTyp (EVar $ fst $ bindingBinder bind) (EVar this) creFs
-      return $ _oUpdate updOut
-    _ -> return emptyUpdateSet
+      subst <- bindCids True updTyp (EVar $ fst $ bindingBinder bind) (EVar this) creFs
+      return (_oUpdate updOut, subst)
+    _ -> return (emptyUpdateSet, emptyExprSubst)
   extVarEnv (fst $ bindingBinder bind)
-  bodyOut <- genExpr False body
-  case _oExpr bodyOut of
-    EUpdate bodyUpd -> do
-      (bodyUpdOut, bodyTyp, creFs) <- genUpdate bodyUpd
-      return ( Output
-                 (_oExpr bodyUpdOut)
-                 (_oUpdate bindOut
-                   `concatUpdateSet` bindUpd
-                   `concatUpdateSet` _oUpdate bodyOut
-                   `concatUpdateSet` _oUpdate bodyUpdOut)
-             , bodyTyp
-             , creFs )
-    _ -> error "Impossible: The body of a bind should be an update expression"
+  bodyOut <- genExpr False $ substituteTm subst body
+  let bodyUpd = case _oExpr bodyOut of
+        EUpdate bodyUpd -> bodyUpd
+        -- Note: This is a bit of a hack, as we're forced to provide some type to
```

Comment:
Why does the type not matter?

Reply:
The only place where these types are used, is in `bindCids`, where they are used to bind fetched / created contract cids. However, as a `UPure` update can't fetch or create a new contract, its type specifically is not important.
But you're right that that's not very clear at all. I added a note.